### PR TITLE
[WIP] chore: move datafusion to sqlglot

### PR DESCRIPTION
**Pull request by mesejo**
This PR moves DataFusion to sqlglot. Kudos to @cpcloud for doing most of the heavy lifting.

### CODE REVIEW

**Comment by cpcloud** [comment]
@mesejo I fixed up a couple issues:

1. made a few window functions work using our new rewrite rules
2. added support for binary literals
3. added support for inet literals

Would you mind cataloging here the missing functionality that was previously passing that may need upstream changes to datafusion to make work?

For example:

1. semi and anti joins work in the python API, but not the SQL API.

File: ibis/backends/datafusion/compiler/values.py [hunk 1]
```diff
@@ -442,9 +442,9 @@ def string_find(op, *, arg, substr, start, end, **_):
 
     if start is not None:
         pos = F.strpos(F.substr(arg, start + 1), substr)
-        return F.coalesce(F.nullif(pos + start, start), 0) - 1
+        return F.coalesce(F.nullif(pos + start, start), 0)
 
-    return F.strpos(arg, substr) - 1
+    return F.strpos(arg, substr)
```
Comment (cpcloud) [inline ibis/backends/datafusion/compiler/values.py:450]: Ah, thank you! I forgot to do this :)

File: ibis/backends/datafusion/compiler/values.py [hunk 2]
```diff
@@ -231,6 +232,8 @@ def _cast(op, *, arg, to, **_):
     if to.is_timestamp() and (timezone := to.timezone) is not None:
         unit = to.unit.name.capitalize()
         return F.arrow_cast(arg, f'Timestamp({unit}, Some("{timezone}"))')
+    if to.is_decimal():
+        return F.arrow_cast(arg, f"{PyArrowType.from_ibis(to)}".capitalize())
```
Comment (cpcloud) [inline ibis/backends/datafusion/compiler/values.py:236]: Nice!

File: ibis/backends/datafusion/compiler/relations.py [hunk 3]
```diff
@@ -40,8 +40,8 @@ def _self_ref(op, *, table, **_):
     ops.LeftJoin: "left",
     ops.RightJoin: "right",
     ops.OuterJoin: "full",
-    ops.LeftAntiJoin: "anti",
-    ops.LeftSemiJoin: "semi",
+    ops.LeftAntiJoin: "left anti",
+    ops.LeftSemiJoin: "left semi",
```
Comment (cpcloud) [inline ibis/backends/datafusion/compiler/relations.py:44]: Wow, what an easy fix!

File: ibis/backends/base/sqlglot/__init__.py [hunk 4]
```diff
@@ -62,12 +62,53 @@ def __getitem__(self, key: str) -> sg.exp.Column:
...
+def parenthesize(op, arg):
+    import ibis.expr.operations as ops
+
+    if isinstance(op, (ops.Binary, ops.Unary)):
+        return paren(arg)
+    # function calls don't need parens
+    return arg
+
+
+def lit(val):
```
Comment (cpcloud) [inline ibis/backends/base/sqlglot/__init__.py:101]: I will clean this up, we don't actually need it: sqlglot has a `convert` function that does essentially the same thing as this function.

File: ibis/backends/datafusion/tests/test_udf.py [hunk 5]
```diff
@@ -75,6 +74,4 @@ def test_builtin_agg_udf_filtered(con):
     def median(a: float, where: bool = True) -> float:
         """Median of a column."""
 
-    expr = median(con.tables.batting.G)
-    with pytest.raises(exc.OperationNotDefinedError, match="No translation rule for"):
-        con.execute(expr)
+    median(con.tables.batting.G)
```
Comment (cpcloud) [inline ibis/backends/datafusion/tests/test_udf.py:77]: Perhaps we should call `.execute()` on this?

**Review by cpcloud — COMMENTED** [review]
Looking great!

A couple of comments.

File: ibis/backends/tests/test_join.py [hunk 6]
```diff
@@ -122,7 +122,7 @@ def test_mutating_join(backend, batting, awards_players, how):
 
 
 @pytest.mark.parametrize("how", ["semi", "anti"])
-@pytest.mark.notimpl(["dask", "druid"])
+@pytest.mark.notimpl(["dask", "druid", "datafusion"])
```
[the remark this answers is not in the document]
Comment (cpcloud) [reply ibis/backends/tests/test_join.py]: This is fixed now :)

File: ibis/backends/datafusion/udfs.py [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+
+import ibis.expr.datatypes as dt  # noqa: TCH001
+
+
+def _extract_epoch_seconds(array) -> dt.int32:
+    return pc.cast(pc.divide(pc.cast(array, pa.int64()), 1_000_000), pa.int32())
+
+
+def extract_epoch_seconds_date(array: dt.date) -> dt.int32:
+    return _extract_epoch_seconds(array)
```
[the remark this answers is not in the document]
Comment (cpcloud) [reply ibis/backends/datafusion/udfs.py:17]: I think we just don't have a test for this yet. That can be done in a follow-up.

File: ibis/backends/base/sqlglot/__init__.py (hunk 4 above)
Comment (cpcloud) [reply ibis/backends/base/sqlglot/__init__.py]: Done!

File: ibis/backends/datafusion/tests/test_udf.py (hunk 5 above)
Comment (cpcloud) [reply ibis/backends/datafusion/tests/test_udf.py]: Done!

**Comment by cpcloud** [comment]
@mesejo This looks great to me!

If you want to move it out of draft mode if you're happy with it, we can merge!

**Review by cpcloud — APPROVED** [review]
Excellent work @mesejo!

**Comment by mesejo** [comment]
> @mesejo This looks great to me!
> 
> If you want to move it out of draft mode if you're happy with it, we can merge!

I also have some other improvements, but I think I'll tackle those in follow-up PRs

**Pull request merged**